src/bom/extract.rs

```diff
@@ -102,45 +83,32 @@ pub fn do_bitcode_extraction(extract_options : &ExtractOptions,
     //
     // NOTE: Ideally, we would be able to use the tar library for this instead
     // of calling out to tar.
-    let mut tar_args = Vec::new();
-    tar_args.push(OsString::from("xif"));
-    tar_args.push(OsString::from(ok_tar_name));
-    match Command::new("tar").args(&tar_args).current_dir(tmp_path).spawn() {
-        Err(msg) => {
-            return Err(anyhow::Error::new(ExtractError::ErrorRunningCommand(String::from("tar"), tar_args, msg)));
-        }
-        Ok(mut child) => {
-            let _rc = child.wait();
-        }
-    }
+
+    extract_ops.push_op(
+        &SubProcOperation::new(
+            &Executable::new("tar",
+                             ExeFileSpec::Append,
+                             ExeFileSpec::NoFileUsed))
+            .set_label("tar unpack")
+            .push_arg("xif")
+            .set_dir(tmp_path)  // Extracted files will be placed here
+            // Input here is explicitly the section dump target file rather
+            // than the output of the previous op.
+            .set_input_file(&FileArg::loc(tar_path))
+    );
 
     // Now all the files extracted from bitcode.tar into tmp_dir should be linked
     // together to create the final bitcode file.
 
-    let mut llvm_link_args = Vec::new();
-    llvm_link_args.push(OsString::from("-o"));
-    llvm_link_args.push(OsString::from(&extract_options.output));
-
-    let mut bc_glob = String::new();
-    bc_glob.push_str(&OsString::from(tmp_path).into_string().unwrap());
-    bc_glob.push_str("/*.bc");
-    let bc_files = glob::glob(&bc_glob)?;
-    for bc_entry in bc_files {
-        let bc_file = bc_entry?;
-        debug!("#: bitcode file: {:?}", bc_file.file_name().unwrap());
-        llvm_link_args.push(OsString::from(bc_file));
-    }
-
-    let llvm_link = OsString::from(extract_options.llvm_link_path.as_ref().unwrap_or(&String::from("llvm-link")));
-    match Command::new(&llvm_link).args(&llvm_link_args).spawn() {
-        Err(msg) => {
-            let llvm_link_str = llvm_link.into_string().unwrap();
-            return Err(anyhow::Error::new(ExtractError::ErrorRunningCommand(llvm_link_str, llvm_link_args, msg)));
-        }
-        Ok(mut child) => {
-            let _rc = child.wait();
-        }
-    }
-
-    Ok(0)
+    let linker = String::from(extract_options.llvm_link_path
+                              .as_ref().unwrap_or(&"llvm-link"
+                                                  .to_string()));
+    let mut link_bc_files = extract_ops.push_op(
+        &SubProcOperation::new(&Executable::new(&linker,
+                                                ExeFileSpec::Append,
+                                                ExeFileSpec::option("-o"))));
+    link_bc_files.set_input_file(&FileArg::glob_in(tmp_path, "*.bc"));
+
+    let executor = get_executor(extract_options.verbose.len());
+    execute_here(&mut extract_ops, &executor).map(|_| 0)
```

Comment:
Why is `execute_here` a free function? Would it make sense for it to be a method on the `ChainedOp`?

Also, what is the idea behind the `map` call? Is that discarding the return code?

Reply:
Good idea.  There were other reasons initially, but there is no reason in its current implementation why it shouldn't be a default method in the Trait.  I've updated chainsop and make the corresponding change here.

Yes, the map call is just discarding the actual return value and returning the expected 0 in the success case.  Let me know if you'd prefer to see another idiom used.